examples/dynamic_watcher.rs

```diff
@@ -28,13 +34,28 @@ async fn main() -> anyhow::Result<()> {
     let api = Api::<DynamicObject>::all_with(client, &ar);
 
     // Fully compatible with kube-runtime
-    let mut items = watcher(api, ListParams::default()).applied_objects().boxed();
-    while let Some(p) = items.try_next().await? {
-        if caps.scope == Scope::Cluster {
-            info!("saw {}", p.name_any());
-        } else {
-            info!("saw {} in {}", p.name_any(), p.namespace().unwrap());
+    if should_watch_meta {
+        let mut items = watcher::watch_metadata(api, ListParams::default())
+            .applied_objects()
+            .boxed();
+
+        while let Some(p) = items.try_next().await? {
+            if caps.scope == Scope::Cluster {
+                info!("saw {}", p.name_any());
+            } else {
+                info!("saw {} in {}", p.name_any(), p.namespace().unwrap());
+            }
         }
-    }
+    } else {
+        let mut items = watcher(api, ListParams::default()).applied_objects().boxed();
+        while let Some(p) = items.try_next().await? {
+            if caps.scope == Scope::Cluster {
+                info!("saw {}", p.name_any());
+            } else {
+                info!("saw {} in {}", p.name_any(), p.namespace().unwrap());
+            }
+        }
```

Comment:
Are these two while loop bodies not equivalent now per watcher?

I feel like you should be able to do 
```rust
let mut items = should_watch_meta {
    watch_metadata()..
} else { watcher().. }.boxed(); // or something like this
while let Some(p) = items.try.next().await? { ... }
```
and use the same loop body given the compatible polymorphic type

Reply:
I tried something similar at first (by doing `.applied_objects()` and then boxing inside each case) but it doesn't work.  Maybe I'm missing something.

```
error[E0308]: `if` and `else` have incompatible types
   --> examples/dynamic_watcher.rs:40:9
  
    = note: expected opaque type `impl Stream<Item = Result<kube::kube_runtime::watcher::Event<PartialObjectMeta>, kube::kube_runtime::watcher::Error>> + std::marker::Send`
               found opaque type `impl Stream<Item = Result<kube::kube_runtime::watcher::Event<DynamicObject>, kube::kube_runtime::watcher::Error>> + std::marker::Send`

```

Since they are both opaque types, I assume the compiler can't infer that they're similar (or there's some polymorphism going around) since the internals of the two functions (`watch` and `metadata_watcher`) may differ (and as such the hidden types might not match). So even if we get a `WatchEvent<PartialObjectMeta>`, and it satisfies the trait bounds (`K: Resource`), it's still not the same as `WatchEvent<K>`. 

Does this line up with either of your understandings?

Reply:
yeah, that makes sense. but the type you are trying to act on is still compatible with a signature that takes an `impl Stream<Item = Result<Event<K>>` where `K: Resource` (or some BoxStream generic over K). not sure what you have to do inline to make that work inline inside `main`, but you could factor it out to a fn that is generic over K to avoid duplicating logic.

Reply:
Gotcha! Was a bit slow here; hard to gauge what and how to change while also keeping the diff relatively small :) Lmk if this works for you.

Reply:
This looks great 👍